File: Arrays/arrays.c
#include <stdio.h>
#include <stdint.h>
#include <time.h>

// Testing: https://www.programiz.com/c-programming/online-compiler/

int main() {
  int arr_size = 2000000;
  int arr[arr_size];
  // Creación de un array de un millón de enteros
  for (int i = 0; i < arr_size; i++) {
    arr[i] = i;
  }

  clock_t start_time = clock();
  for (int i = 0; i < arr_size; i++) {
    /*
    `volatile: Es un calificador de tipo que le dice al compilador que 
    una variable puede cambiar en cualquier momento sin ninguna acción 
    aparente en el código que el compilador encuentra.
    */
    volatile int x  = arr[i];
  }

  // Convertir a milisegundos
  double elapsed_time = ((double)(clock() - start_time) / CLOCKS_PER_SEC) * 1000;  

  printf("                 ,\n");
  printf("                /|      __\n");
  printf("               / |   ,-~ /\n");
  printf("              Y :|  //  /\n");
  printf("              | jj /( .^\n");
  printf("              >-\"~\"-v\"\n");
  printf("             /       Y\n");
  printf("            jo  o    |\n");
  printf("           ( ~T~     j\n");
  printf("            >._-' _./\n");
  printf("           /   \"~\"  |\n");
  printf("          Y     _,  |\n");
  printf("         /| ;-\"~ _  l\n");
  printf("        / l/ ,-\"~    \\\n");
  printf("        \\//\\/      .- \\\n");
  printf("         Y        /    Y\n");
  printf("         l       I     !\n");
  printf("         ]\\      _\\    /\"\\\n");
  printf("        (\" ~----( ~   Y.  )\n");
  printf("╔════════════════════════════════╗\n");
  printf("║         Arrays en C           ║\n");
  printf("║--------------------------------║\n");
  printf("║ Tiempo de compilación: %.2f ms ║\n", elapsed_time);
  printf("╚════════════════════════════════╝\n");

  return 0;
}

File: Arrays/arrays.rs
use std::time::Instant;

// Testing: https://www.programiz.com/rust/online-compiler/

fn main() {
  // Creación de un array de un millón de enteros
  let arr: [i32; 2000000] = [0; 2000000];
  for (i, elem) in arr.iter().enumerate() {
    let _ = elem + i as i32;
  }

  let start_time = Instant::now();
  for i in 0..2000000 {
    let _ = arr[i];
  }
  
  // Convertir a milisegundos
  let elapsed_time = start_time.elapsed().as_millis();

  println!("            _______  ");
  println!("          _/       \\_");
  println!("         / |       | \\");
  println!("        /  |__   __|  \\");
  println!("       |__/((o| |o))\\__|");
  println!("       |      | |      |");
  println!("       |\\     |_|     /|");
  println!("       | \\           / |");
  println!("        \\| /  ___  \\ |/");
  println!("         \\ | / _ \\ | /");
  println!("          \\_________/");
  println!("           _|_____|_");
  println!("      ____|_________|____");
  println!("     /                   \\"); 
  println!("╔═════════════════════════════╗");
  println!("║     Arrays en Rust 🦀       ║");
  println!("║-----------------------------║");
  println!("║ Tiempo de compilación: {:.2}ms ║", elapsed_time);
  println!("╚═════════════════════════════╝");

}
